### ISSUE TITLE
Proposal Feedback

### ISSUE THREAD

**Issue by ngoldbaum** [AshKelly/gsoc-2018#1]
Thanks so much for going through the GSOC process and giving me an opportunity to look over your proposal.

I think this is a very strong proposal. I have a few general comments below but overall I think this is almost ready to submit.

I'm pinging @langmm, @matthewturk, as well as @chummels and @dnarayanan  to look this over as well if they have time before the deadline (March 27). I know Cameron and Desika aren't listed as  mentors but I bet they're interested in this proposal.

## Community bonding period

I don't see what your plans during this period are. Can you add some details? Here's some background on what exactly this is:

http://write.flossmanuals.net/gsocstudentguide/how-to-get-a-head-start/

Ideally you'll be 100% ready to start working and interacting with the community on day 1.

## Documentation focus

I'm ashamed to admit that the documentation has largely not yet been updated for the demeshening branch. As I'm the one who has led development so far, that's mostly my fault. That said, I'm glad you've identified that the documentation is lacking and that you plan to spend time writing new docs as you familiarize yourself with how yt works in the demeshening branch. I think these docs will be very valuable. I think contributing to a project's documentation is also a great way to learn a project - you're spending your time working on the documentation mostly reading documentation after all :)

## Octree interpolation

Right now this is listed as a stretch goal. Instead, I'd like to see it listed in the deliverables. Personally, I think this is more important to finish than fields that need spatial derivatives. It's also harder, of course and I very much understand the desire to keep the proposal scope manageable. However, I want to encourage you to think about including it in the proposal since I think it's a very desirable feature, especially from current users who need to port old scripts.

This subproject would require creating a new yt data object after you get familiar with the internals of other yt data objects via `YTCoveringGrid` and `YTArbitraryGrid`. You'd be creating something like `YTParticleOctree` that would have an API like this:

```python
left_edge = [.2, .3, .4]
right_edge = [1, 1, 1]
number_to_refine_by = 64
otree = ds.particle_octree(left_edge, right_edge, n_ref, refine_field='particle_mass')
density_values = octree['density']
```

This will require writing an in-memory octree data structure in cython, and then code to deposit or SPH smooth data onto that octree based on the position of the data. Most of the high-level implementation will follow the existing data objects that exist large to deposit or smooth data onto. Ideally you'd be able to reuse the utility functions you'll write for the uniform resolution grids with the octree. For example, for gather smoothing you'd have a utility function that consumes chunks of a uniform resolution grid or leaf octs, and iterates over the zones in the chunk or oct to find nearest-neighbor lists and smooth data. 

I think in the context of a full-time GSOC commitment this is doable in a couple of weeks. If things go smoothly you might also be able to get ahead of your current estimates for finishing uniform resolution smoothing and KDTree search utilities. Of course estimates for how long it will take to develop software are bunk and my estimate may be bogus. If we get to a point where you feel like the octree deposition subproject will be too much or it will bog you down then we will have a way for you to voice that concern and allow us to agree to abandon it.

One way to structure this uncertainty in your proposal is to include a few events in your project timeline when we meet and evaluate whether we should abandon any deliverables or rework the timeline. You could also evaluate the risk level of your deliverables now and we can reassess after you've gotten more familiar with yt and have made some progress.

## Algorithms

It might be worthwhile to dive into the literature a bit to see if we can figure out ways to optimize and parallelize KDTree searching. For example, as you iterate over a chunk of data, the kdtree search for each zone in the chunk will be mostly redundant. One speedup would be to cache the last search path through the KDTree and then merely correct that search path. KDTrees have a significant presence in algorithms and CS literature so it would probably be worthwhile for you to spend some time in the community bonding period doing some literature searching and reading papers to try to find algorithms we can use.

For parallelization, I tried to get the KDTree search to work with cython's OpenMP wrappers but ran into issues which might be due to cython bugs - we could try to resurrect that work. It would be really awesome (especially on modern workstations with dozens of cores) if we could make the KDTree search exploit parallelism as much as possible. This is something that you could spend a day or two working on while you are writing KDTree search utility functions to see what can be done.

**Comment by ngoldbaum** [comment]
Ping @JBorrow too.

**Comment by JBorrow** [comment]
I agree with pretty much all of @ngoldbaum's comments; it's probably worth noting that for NN searching we could probably help out locally (via the SWIFT team) here in Durham -- we have put a lot of effort into developing scaling search algorithms (see e.g. https://arxiv.org/pdf/1309.3783.pdf).

**Comment by MatthewTurk** [comment]
I agree with @ngoldbaum as well.  Thank you for your work on this proposal!
